File: frontend/src/types/vendorDependency.ts
export type VendorRelationshipType = 'subcontractor' | 'reseller' | 'parent_company' | 'other';

export interface VendorRelationship {
    id: number;
    vendor_id: number;
    related_vendor_id: number;
    related_vendor_name?: string | null;
    relationship_type: VendorRelationshipType;
    created_at: string;
}

export interface VendorDependency {
    id: number;
    vendor_service_id: number;
    risk_id?: number | null;
    risk_name?: string | null;
    department_id?: number | null;
    department_name?: string | null;
    supported_function_name?: string | null;
    created_at: string;
}

export interface VendorService {
    id: number;
    vendor_id: number;
    service_name: string;
    notes?: string | null;
    dependencies: VendorDependency[];
    created_at: string;
    updated_at: string;
}

export interface VendorDependencyGraphNode {
    vendor_id: number;
    vendor_name: string;
    relationship_type?: VendorRelationshipType | null;
    children: VendorDependencyGraphNode[];
}

export interface VendorConcentrationFlag {
    key: string;
    severity: string;
    reason: string;
}

export interface VendorConcentrationSummary {
    score: number;
    flags: VendorConcentrationFlag[];
}

export interface VendorDependenciesResponse {
    vendor_id: number;
    relationships: VendorRelationship[];
    services: VendorService[];
    relationship_tree: VendorDependencyGraphNode;
    concentration: VendorConcentrationSummary;
}

export interface VendorRelationshipCreate {
    related_vendor_id: number;
    relationship_type: VendorRelationshipType;
}

export interface VendorServiceCreate {
    service_name: string;
    notes?: string | null;
}

export interface VendorDependencyCreate {
    risk_id?: number | null;
    department_id?: number | null;
    supported_function_name?: string | null;
}

File: backend/app/schemas/vendor_dependency.py
from __future__ import annotations

from enum import Enum
from datetime import datetime

from pydantic import BaseModel, Field


class VendorRelationshipTypeEnum(str, Enum):
    subcontractor = "subcontractor"
    reseller = "reseller"
    parent_company = "parent_company"
    other = "other"


class VendorRelationshipRead(BaseModel):
    id: int
    vendor_id: int
    related_vendor_id: int
    related_vendor_name: str | None = None
    relationship_type: VendorRelationshipTypeEnum
    created_at: datetime


class VendorRelationshipCreate(BaseModel):
    related_vendor_id: int
    relationship_type: VendorRelationshipTypeEnum = VendorRelationshipTypeEnum.subcontractor


class VendorDependencyRead(BaseModel):
    id: int
    vendor_service_id: int
    risk_id: int | None = None
    risk_name: str | None = None
    department_id: int | None = None
    department_name: str | None = None
    supported_function_name: str | None = None
    created_at: datetime


class VendorDependencyCreate(BaseModel):
    risk_id: int | None = None
    department_id: int | None = None
    supported_function_name: str | None = Field(None, max_length=255)


class VendorServiceRead(BaseModel):
    id: int
    vendor_id: int
    service_name: str
    notes: str | None = None
    dependencies: list[VendorDependencyRead]
    created_at: datetime
    updated_at: datetime


class VendorServiceCreate(BaseModel):
    service_name: str = Field(..., max_length=255)
    notes: str | None = None


class VendorServiceUpdate(BaseModel):
    service_name: str | None = Field(None, max_length=255)
    notes: str | None = None


class VendorDependencyGraphNode(BaseModel):
    vendor_id: int
    vendor_name: str
    relationship_type: VendorRelationshipTypeEnum | None = None
    children: list["VendorDependencyGraphNode"] = Field(default_factory=list)


VendorDependencyGraphNode.model_rebuild()


class VendorConcentrationFlag(BaseModel):
    key: str
    severity: str
    reason: str


class VendorConcentrationSummary(BaseModel):
    score: int
    flags: list[VendorConcentrationFlag]


class VendorDependenciesResponse(BaseModel):
    vendor_id: int
    relationships: list[VendorRelationshipRead]
    services: list[VendorServiceRead]
    relationship_tree: VendorDependencyGraphNode
    concentration: VendorConcentrationSummary
